internal/expr/prepare.go

```diff
@@ -56,28 +56,139 @@ func starCount(fns []fullName) int {
 }
 
 // prepareInput checks that the input expression corresponds to a known type.
-func prepareInput(ti typeNameToInfo, p *inputPart) (typeMember, error) {
-	info, ok := ti[p.sourceType.prefix]
-	if !ok {
-		ts := getKeys(ti)
-		if len(ts) == 0 {
-			return nil, fmt.Errorf(`type %q not passed as a parameter`, p.sourceType.prefix)
-		} else {
-			return nil, fmt.Errorf(`type %q not passed as a parameter, have: %s`, p.sourceType.prefix, strings.Join(ts, ", "))
+func prepareInput(ti typeNameToInfo, p *inputPart) ([]fullName, []typeMember, error) {
+	var inCols = make([]fullName, 0)
+	var typeMembers = make([]typeMember, 0)
+
+	numTypes := len(p.sourceTypes)
+	numColumns := len(p.targetColumns)
+	starTypes := starCount(p.sourceTypes)
+	starColumns := starCount(p.targetColumns)
+
+	// Check target struct type and its tags are valid.
+	var info typeInfo
+	var ok bool
+	var err error
```

Comment:
I might be missing something, but I can't seem to see where these are used? All the places I see have them shadowed, so would it be better to remove these entirely? Otherwise, could they be declared closer to where they're used?

Reply:
I've moved them to be closer to where they are used. Having them here makes the code shorter, but redeclaring them where they are used makes it more explicit.